layout errors in transmute checking don't get emitted

https://github.com/rust-lang/rust/blob/672388edbee9e93c35e5fdf7dac818a6612a5103/tests/ui/layout/normalization-failure.rs
```rust
use std::marker::PhantomData;

trait Project1 {
    type Assoc1;
}

impl<T> Project1 for T {
    type Assoc1 = ();
}

trait Project2 {
    type Assoc2;
    fn get(self) -> Self::Assoc2;
}

impl<T: Project1<Assoc1 = ()>> Project2 for PhantomData<T> {
    type Assoc2 = ();
    fn get(self) -> Self::Assoc2 {}
}

fn opaque<T>() -> impl Project2 {
    PhantomData::<T>
}

fn check<T: Project1>() {
    unsafe {
        std::mem::transmute::<_, ()>(opaque::<T>().get());
        //~^ ERROR: cannot transmute
        //~| NOTE: source type: `{type error}` (the type has an unknown layout)
        //~| NOTE: target type: `()` (0 bits)
    }
}
```
results in
```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> src/lib.rs:28:9
   |
28 |         std::mem::transmute::<_, ()>(opaque::<T>().get());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `{type error}` (the type has an unknown layout)
   = note: target type: `()` (0 bits)

For more information about this error, try `rustc --explain E0512`.
```
This error message is weird:
- it references `type error` even though no other error has been emitted
- we don't explain why the source error is a type error
- 
We should have some `LayoutError::NormalizationFailure` here which should also be emitted and then don't emit the transmute error as one of the types is a `type error`

--- a/compiler/rustc_hir_typeck/src/intrinsicck.rs
+++ b/compiler/rustc_hir_typeck/src/intrinsicck.rs
@@ -75,20 +75,20 @@ fn check_transmute<'tcx>(
     hir_id: HirId,
 ) -> Result<(), ErrorGuaranteed> {
     let span = tcx.hir_span(hir_id);
-    let normalize = |ty| {
-        if let Ok(ty) = tcx.try_normalize_erasing_regions(typing_env, ty) {
-            ty
-        } else {
-            Ty::new_error_with_message(
-                tcx,
-                span,
-                format!("tried to normalize non-wf type {ty:#?} in check_transmute"),
-            )
-        }
+    let normalize = |ty: Unnormalized<'tcx, Ty<'tcx>>| -> Result<Ty<'tcx>, ErrorGuaranteed> {
+        let unnormalized_ty = ty.skip_normalization();
+        tcx.try_normalize_erasing_regions(typing_env, ty).map_err(|err| {
+            tcx.dcx()
+                .struct_span_err(
+                    span,
+                    LayoutError::NormalizationFailure(unnormalized_ty, err).to_string(),
+                )
+                .emit()
+        })
     };
 
-    let from = normalize(from);
-    let to = normalize(to);
+    let from = normalize(from)?;
+    let to = normalize(to)?;
     trace!(?from, ?to);
 
     // Transmutes that are only changing lifetimes are always ok.

--- a/compiler/rustc_hir_typeck/src/intrinsicck.rs
+++ b/compiler/rustc_hir_typeck/src/intrinsicck.rs
@@ -75,20 +75,15 @@ fn check_transmute<'tcx>(
     hir_id: HirId,
 ) -> Result<(), ErrorGuaranteed> {
     let span = tcx.hir_span(hir_id);
-    let normalize = |ty| {
-        if let Ok(ty) = tcx.try_normalize_erasing_regions(typing_env, ty) {
-            ty
-        } else {
-            Ty::new_error_with_message(
-                tcx,
-                span,
-                format!("tried to normalize non-wf type {ty:#?} in check_transmute"),
-            )
-        }
+    let normalize = |ty: Unnormalized<'tcx, Ty<'tcx>>| -> Result<Ty<'tcx>, ErrorGuaranteed> {
+        tcx.try_normalize_erasing_regions(typing_env, ty).map_err(|err| {
+            let err = LayoutError::NormalizationFailure(ty.skip_normalization(), err);
+            tcx.dcx().struct_span_err(span, err.to_string()).emit()
+        })
     };
 
-    let from = normalize(from);
-    let to = normalize(to);
+    let from = normalize(from)?;
+    let to = normalize(to)?;
     trace!(?from, ?to);
 
     // Transmutes that are only changing lifetimes are always ok.

--- a/tests/ui/layout/normalization-failure.rs
+++ b/tests/ui/layout/normalization-failure.rs
@@ -48,9 +48,7 @@ fn opaque<T>() -> impl Project2 {
 fn check<T: Project1>() {
     unsafe {
         std::mem::transmute::<_, ()>(opaque::<T>().get());
-        //~^ ERROR: cannot transmute
-        //~| NOTE: source type: `{type error}` (the type has an unknown layout)
-        //~| NOTE: target type: `()` (0 bits)
+        //~^ ERROR: unable to determine layout
     }
 }
 

--- a/tests/ui/layout/normalization-failure.stderr
+++ b/tests/ui/layout/normalization-failure.stderr
@@ -1,12 +1,8 @@
-error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
+error: unable to determine layout for `<impl Project2 as Project2>::Assoc2` because `<impl Project2 as Project2>::Assoc2` cannot be normalized
   --> $DIR/normalization-failure.rs:50:9
    |
 LL |         std::mem::transmute::<_, ()>(opaque::<T>().get());
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-   |
-   = note: source type: `{type error}` (the type has an unknown layout)
-   = note: target type: `()` (0 bits)
 
 error: aborting due to 1 previous error
 
-For more information about this error, try `rustc --explain E0512`.

